### DISCUSSION TITLE
SilverLining / Rocky integration ideas

### DISCUSSION THREAD

**Discussion by fkane** [pelicanmapping/rocky#22] [Q&A / Support]
We've got a customer asking about integrating [SilverLining's new VSG support](https://sundog-soft.com/2023/12/silverlining-brings-3d-clouds-and-skies-to-vulkanscenegraph/) with Rocky for a pretty cool commercial application. Just wanted to check on any high-level guidance you might have for this sort of thing!

We've successfully wrapped SilverLining's sky and clouds in a vsg::Group, so in theory it's a matter of finding the top-level group for Rocky's scene and adding it as a child. Then we'd have to do stuff to sync up SilverLining's latitude, longitude, time of day, and "up" vector with Rocky like you're doing with osgEarth, along with a few changes to our reference VSG integration to put SilverLining into geocentric mode.

I can't build Rocky at the moment and play with it due to [this issue](https://github.com/vsg-dev/VulkanSceneGraph/issues/1035), but that's not your fault. But just so I can start wrapping my head around it, is that top level group exposed to the application somehow, as well as the current camera position? Or would we need to go the osgEarth route of integrating at a lower level?

**Comment by gwaldron** [comment]
Frank, the Rocky API is very open, like VSG. Most members are public and it's easy to get into the guts of things.

The `Application` object has a scene graph that starts with `root` group and has one child called `mainScene`. The easiest move (at least for initial testing) be to put your group in between these two.

To get the coordinates of the current focal point of a view's camera, you can call
```c++
auto manip = app.displayManager->getMapManipulator(view);
auto viewpoint = manip->getViewpoint();
```
And that viewpoint will contains a `point` with the coordinate.

I hope that works :) and I am happy to assist with testing.

**Reply by fkane** [reply]
Excellent! Once I can get things building I'll have a go at it. Thanks.

**Reply by fkane** [reply]
Yay, got everything working!

![rocky-silverlining](https://github.com/pelicanmapping/rocky/assets/2250566/85f13c57-c4ae-49e0-9dad-ace8a4742961)
![rocky-earth](https://github.com/pelicanmapping/rocky/assets/2250566/53e58a23-a53c-4636-8f92-f8d224ed89a9)

Turned out the right place to insert SilverLining was at the end of the mainScene's children; it has translucent stuff (clouds) so it should draw last:

```
    auto silverLiningGroup = SilverLiningGroup::create(window, app);
    auto silverLiningCommand = SilverLiningCommand::create(silverLiningGroup->atmosphere(), silverLiningGroup);
    silverLiningGroup->addChild(silverLiningCommand);
    app.mainScene->children.insert(app.mainScene->children.end(), silverLiningGroup);
```

I didn't find a getMapManipulator function, but it turned out we needed the actual camera position and not the focal point anyhow. I ended up getting that in a somewhat convoluted way, but it works:

```
    auto first_view = _rockyApp.displayManager->windowsAndViews.begin()->second.front();
    if (first_view)
    {
        auto viewmat = first_view->camera->viewMatrix;
        auto lookAt = dynamic_cast<vsg::LookAt*>(viewmat.get());
        if (lookAt) {
            auto eyePos = lookAt->eye;
            rocky::GeoPoint geoPoint(rocky::SRS::ECEF, eyePos);
```

We included our sample Rocky integration (based on the rdemo app) in our SilverLining 7 release, with copious warnings about the Rocky API being early and fluid. But it's one step closer to giving the simulation & training world a migration path to Vulkan.

**Reply by gwaldron** [reply]
This is great. Thanks Frank :)
